crates/circuit/src/parameter_table.rs

```diff
@@ -10,164 +10,395 @@
+use crate::imports::UUID;
 
-use hashbrown::{HashMap, HashSet};
+import_exception!(qiskit.circuit, CircuitError);
+
+#[derive(Error, Debug)]
+pub enum ParameterTableError {
+    #[error("parameter '{0:?}' is not tracked in the table")]
+    ParameterNotTracked(ParameterUuid),
+    #[error("usage {0:?} is not tracked by the table")]
+    UsageNotTracked(ParameterUse),
+}
+impl From<ParameterTableError> for PyErr {
+    fn from(value: ParameterTableError) -> PyErr {
+        CircuitError::new_err(value.to_string())
+    }
+}
+
+/// A single use of a symbolic parameter.
+#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+pub enum ParameterUse {
+    Index { instruction: usize, parameter: u32 },
+    GlobalPhase,
+}
+
+/// Rust-space extra information that a `ParameterVectorElement` has.  This is used most heavily
+/// during sorting; vector elements are sorted by their parent name, and index within that.
+#[derive(Clone, Debug)]
+struct VectorElement {
+    vector_uuid: VectorUuid,
+    index: usize,
+}
+
+/// Tracked data tied to each parameter's UUID in the table.
+#[derive(Clone, Debug)]
+pub struct ParameterInfo {
+    uses: HashSet<ParameterUse>,
+    name: PyBackedStr,
+    element: Option<VectorElement>,
+    object: Py<PyAny>,
+}
 
-/// The index value in a `ParamEntry` that indicates the global phase.
-pub const GLOBAL_PHASE_INDEX: usize = usize::MAX;
+/// Rust-space information on a Python `ParameterVector` and its uses in the table.
+#[derive(Clone, Debug)]
+struct VectorInfo {
+    name: PyBackedStr,
+    /// Number of elements of the vector tracked within the parameter table.
+    refcount: usize,
+}
 
-#[pyclass(freelist = 20, module = "qiskit._accelerate.circuit")]
-pub(crate) struct ParamEntryKeys {
-    keys: Vec<(usize, usize)>,
-    iter_pos: usize,
+/// Type-safe UUID for a symbolic parameter.  This does not track the name of the `Parameter`; it
+/// can't be used alone to reconstruct a Python instance.  That tracking remains only withing the
+/// `ParameterTable`.
+#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+pub struct ParameterUuid(u128);
+impl ParameterUuid {
+    /// Extract a UUID from a Python-space `Parameter` object. This assumes that the object is known
+    /// to be a parameter.
+    pub fn from_parameter(ob: &Bound<PyAny>) -> PyResult<Self> {
+        ob.getattr(intern!(ob.py(), "_uuid"))?.extract()
+    }
 }
 
-#[pymethods]
-impl ParamEntryKeys {
-    fn __iter__(slf: PyRef<Self>) -> Py<ParamEntryKeys> {
-        slf.into()
+/// This implementation of `FromPyObject` is for the UUID itself, which is what the `ParameterUuid`
+/// struct actually encapsulates.
+impl<'py> FromPyObject<'py> for ParameterUuid {
+    fn extract_bound(ob: &Bound<'py, PyAny>) -> PyResult<Self> {
+        if ob.is_exact_instance(UUID.get_bound(ob.py())) {
+            ob.getattr(intern!(ob.py(), "int"))?.extract().map(Self)
+        } else {
+            Err(PyTypeError::new_err("not a UUID"))
+        }
```

Comment:
We might want to make a note for the future that pyo3 has an open PR adding uuid conversion support for:

https://docs.rs/uuid/latest/uuid/

that might be something we want to leverage in the future here.